src/vocabulary.rs

```diff
@@ -0,0 +1,519 @@
+        NamespacedKeyword::new("db.schema", "attribute")
+    };
+    static ref DB_SCHEMA_VERSION: NamespacedKeyword = {
+        NamespacedKeyword::new("db.schema", "version")
+    };
+    static ref DB_IDENT: NamespacedKeyword = {
+        NamespacedKeyword::new("db", "ident")
+    };
+    static ref DB_UNIQUE: NamespacedKeyword = {
+        NamespacedKeyword::new("db", "unique")
+    };
+    static ref DB_UNIQUE_VALUE: NamespacedKeyword = {
+        NamespacedKeyword::new("db.unique", "value")
+    };
+    static ref DB_UNIQUE_IDENTITY: NamespacedKeyword = {
+        NamespacedKeyword::new("db.unique", "identity")
+    };
+    static ref DB_IS_COMPONENT: NamespacedKeyword = {
+        NamespacedKeyword::new("db", "isComponent")
+    };
+    static ref DB_VALUE_TYPE: NamespacedKeyword = {
+        NamespacedKeyword::new("db", "valueType")
+    };
+    static ref DB_INDEX: NamespacedKeyword = {
+        NamespacedKeyword::new("db", "index")
+    };
+    static ref DB_FULLTEXT: NamespacedKeyword = {
+        NamespacedKeyword::new("db", "fulltext")
+    };
+    static ref DB_CARDINALITY: NamespacedKeyword = {
+        NamespacedKeyword::new("db", "cardinality")
+    };
+    static ref DB_CARDINALITY_ONE: NamespacedKeyword = {
+        NamespacedKeyword::new("db.cardinality", "one")
+    };
+    static ref DB_CARDINALITY_MANY: NamespacedKeyword = {
+        NamespacedKeyword::new("db.cardinality", "many")
+    };
+
+    // Not yet supported.
+    // static ref DB_NO_HISTORY: NamespacedKeyword = {
+    //     NamespacedKeyword::new("db", "noHistory")
+    // };
+}
+
+trait HasCoreSchema {
+    /// Return the entity ID for a type. On failure, return `MissingCoreVocabulary`.
+    fn core_type(&self, t: ValueType) -> Result<KnownEntid>;
+
+    /// Return the entity ID for an ident. On failure, return `MissingCoreVocabulary`.
+    fn core_entid(&self, ident: &NamespacedKeyword) -> Result<KnownEntid>;
+
+    /// Return the entity ID for an attribute's keyword. On failure, return
+    /// `MissingCoreVocabulary`.
+    fn core_attribute(&self, ident: &NamespacedKeyword) -> Result<KnownEntid>;
+}
+
+impl<T> HasCoreSchema for T where T: HasSchema {
+    fn core_type(&self, t: ValueType) -> Result<KnownEntid> {
+        self.entid_for_type(t)
+            .ok_or_else(|| ErrorKind::MissingCoreVocabulary(DB_SCHEMA_VERSION.clone()).into())
+    }
+
+    fn core_entid(&self, ident: &NamespacedKeyword) -> Result<KnownEntid> {
+        self.get_entid(ident)
+            .ok_or_else(|| ErrorKind::MissingCoreVocabulary(DB_SCHEMA_VERSION.clone()).into())
+    }
+
+    fn core_attribute(&self, ident: &NamespacedKeyword) -> Result<KnownEntid> {
+        self.attribute_for_ident(ident)
+            .ok_or_else(|| ErrorKind::MissingCoreVocabulary(DB_SCHEMA_VERSION.clone()).into())
+            .map(|(_, e)| e)
+    }
+}
+
+impl Definition {
+    fn description_for_attributes<'s, T, R>(&'s self, attributes: &[R], via: &T) -> Result<Terms>
```

Comment:
Compare and contrast this to the schema exporting EDN thing you wrote in the earlier patches.  Can you unify?  Can you use this throughout and save some EDN construction?

Reply:
I don't think there's a lot of overlap between `Attribute::to_edn_value` and this, unless we were to write an EDN outputter for `Terms`.